pkg/stanza/fileconsumer/internal/reader/reader.go

```diff
@@ -34,11 +35,13 @@ type Reader struct {
 	set                    component.TelemetrySettings
 	fileName               string
 	file                   *os.File
+	reader                 io.Reader
 	fingerprintSize        int
 	initialBufferSize      int
 	maxLogSize             int
 	lineSplitFunc          bufio.SplitFunc
 	splitFunc              bufio.SplitFunc
+	deferFunc              func()
```

Comment:
It's not clear to me that we actually need this. None of the tests fail if I remove it. Can you add a test that demonstrates why it's necessary?

Reply:
sure, will look into adding a test for this. I'll keep you posted

Reply:
alright, i have added such a test - the issue was appearing in the next`ReadToEnd` iteration after the first time the reader scanned through the whole file. The offset override at the end is necessary to set the reader offset to the end of the compressed file, rather than the uncompressed content (the reader does that at several places within the `ReadToEnd` method. So if the compressed file has a size of e.g. 10B, and the uncompressed content has 100B, the `Offset` of the reader would be `100` at the end of `ReadToEnd`. In the next iteration this would then lead to the `invalid header` error, as the reader, before uncompressing would try to start at index 100, rather than 10. This case slipped through the tests by accident, but by adding another invocation of the `poll()` method the test is failing without that override via the `deferFunc`

Reply:
Thanks for adding this to the test. I see that we definitely need a way to set the offset differently.

This implementation introduces an assumption that `ReadToEnd` should only be used once per `Reader` (because the defered offset override is fixed at the time of Reader creation). I know I suggested consolidating the gzip reader creation logic into the factory, but now I see why we need to create the gzip reader every time `ReadToEnd` is called. I played around with this some more and have [another PR to your branch](https://github.com/bacherfl/opentelemetry-collector-contrib/pull/2).

What I'd _really_ like to figure out is how we could abstract offset management in a way where multiple completely separate implementations are possible, rather than having a per-token implementation that we always use, and an override that sometimes is used. However, I think this may be quite tricky so would be OK with not doing it in this PR.

Reply:
Thanks for your PR - and good thought with moving the reader creation into `ReadToEnd`. I was not considering that that `ReadToEnd` could potentially be called multiple times for the same reader instance, so that fixes this nicely.

Regarding offset management, I agree that this is something we could potentially improve. As of right now, I do not have a clear idea of how to do this optimally. So if this is OK with you, I can look into what we can do here in a separate PR - which would also be helpful for me to get to know this component a bit better and dive further into its details

Reply:
Sounds good. Thanks for contributing this!